File: 조준장/level2/주차 요금 계산.py
import math

def solution(fees, records):
    car = {}
    res_car = {}
    answer_car = {}

    for r in records:
        coin = 0
        time, number, direction = r.split(" ")
        hour, minute = time.split(":")
        total_minute = (int(hour) * 60) + int(minute)
        
        if direction == "IN":
            car[number] = total_minute
        else:
            time_difference = total_minute - car[number]
            car[number] = -1
            if number not in res_car:
                res_car[number] = time_difference
            else:
                res_car[number] += time_difference

    for key, value in car.items():
        if value != -1:
            if key not in res_car:
                res_car[key] = (23*60) + 59 - value
            else:
                res_car[key] += (23*60) + 59 - value
                
    for key, value in res_car.items():
        if value <= fees[0]:
            answer_car[key] = fees[1]
            continue
        coin = fees[1]
        value -= fees[0]
        coin += math.ceil(value / fees[2]) * fees[3]
        answer_car[key] = coin
    
    return [s[1] for s in sorted(answer_car.items(), key=lambda x: x [0])]

File: 조준장/level2/주차 요금 계산.kt
class Solution {
    fun solution(
        fees: IntArray,
        records: Array<String>,
    ): IntArray {
        var answer: IntArray = intArrayOf()
        val resultMap =
            records
                .map { it.split(" ") }
                .groupBy({ it[1] }, { it[0] }).toMutableMap()

        resultMap.forEach {
            if (it.value.size % 2 != 0) {
                resultMap[it.key] = resultMap[it.key]!! + "23:59"
            }
        }

        val answerMap =
            resultMap.entries.map { (key, value) ->
                var time = 0
                for (i in 0..value.size - 2 step 2) {
                    time += value[i + 1].toTime() - value[i].toTime()
                }
                key to calculate(fees, time)
            }

        val sortAnswerMap = answerMap.sortedBy { it.first }.map { it.second }

        return sortAnswerMap.toIntArray()
    }

    fun calculate(
        fees: IntArray,
        time: Int,
    ): Int {
        var nTime = time
        var coin = 0

        if (nTime <= fees[0]) return fees[1]
        coin += fees[1]
        nTime -= fees[0]

        var m = nTime / fees[2]
        val n = nTime % fees[2]
        
        if (n != 0) m++

        coin += fees[3] * m
        return coin
    }

    fun String.toTime(): Int {
        val (hours, min) = this.split(":")
        return hours.toInt() * 60 + min.toInt()
    }
}
